aspnetcore/grpc/services.md

```diff
@@ -0,0 +1,236 @@
+
+```csharp
+public override async Task StreamingFromServer(SimpleRequest request,
+    IServerStreamWriter<SimpleResponse> responseStream, ServerCallContext context)
+{
+    for (var i = 0; i < 5; i++)
+    {
+        await responseStream.WriteAsync(new SimpleResponse());
+        await Task.Delay(TimeSpan.FromSeconds(1));
+    }
+}
+```
+
+The client has no way to send additional messages or data once the server streaming method has started. Some streaming methods are designed to run forever. In this situation a client can cancel the call when they no longer need it. When cancellation happens the client sends a signal to the server and the `ServerCallContext.CancellationToken` <xref:System.Threading.CancellationToken> is raised. The token should be used on the server with async methods used so that any asynchronous work is canceled together with the streaming call, and the method exits quickly.
+
+```csharp
+public override async Task StreamingFromServer(SimpleRequest request,
+    IServerStreamWriter<SimpleResponse> responseStream, ServerCallContext context)
+{
+    while (!context.CancellationToken.IsCancellationRequested)
+    {
+        await responseStream.WriteAsync(new SimpleResponse());
+        await Task.Delay(TimeSpan.FromSeconds(1), context.CancellationToken);
+    }
+}
+```
+
+### Client streaming method
+
+A client streaming method starts *without* the method receiving a message. The `requestStream` parameter is used to read messages from the client. A client streaming call is complete when a response message is returned.
+
+```csharp
+public override async Task<SimpleResponse> StreamingFromClient(
+    IAsyncStreamReader<SimpleRequest> requestStream, ServerCallContext context)
+{
+    while (await requestStream.MoveNext())
+    {
+        var message = requestStream.Current;
+        // ...
+    }
+    return new SimpleResponse();
+}
+```
+
+When using C# 8 or later, the `await foreach` syntax can be used to read messages. The `IAsyncStreamReader<T>.ReadAllAsync()` extension method reads all messages from the request stream:
+
+```csharp
+public override async Task<SimpleResponse> StreamingFromClient(
+    IAsyncStreamReader<SimpleRequest> requestStream, ServerCallContext context)
+{
+    await foreach (var message in requestStream.ReadAllAsync())
+    {
+        // ...
+    }
+    return new SimpleResponse();
+}
+```
+
+### Bi-directional streaming method
+
+A bi-directional streaming method starts *without* the method receiving a message. The `requestStream` parameter is used to read messages from the client. The method can choose to send messages with `responseStream.WriteAsync`. A bi-directional streaming call is complete when the the method returns.
+
+```csharp
+public override async Task ReadySetGo(IAsyncStreamReader<SimpleRequest> requestStream,
```

Comment:
This example is a little simplistic right? We should probably mention that bi-directional stream often involves dispatching the reading of the request stream like the real [RacerService](https://github.com/grpc/grpc-dotnet/blob/master/examples/Racer/Server/Services/RacerService.cs#L36). It might be a little advanced for this doc but we should at least mention it. The current sample only illustrate a ping-pong pattern which doesn't fully illustrate the capabilities of bi-directional streams.